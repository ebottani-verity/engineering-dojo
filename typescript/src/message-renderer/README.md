# Concept - Four pillars of a good test
Vladimir Khorikov describes these pillars of a good test in chapter 4 of his book “Unit Testing Principles, Practices and Patterns”:

Protection against regressions
Resistance to refactoring
Fast feedback
Maintainability

# MessageRenderer Test Design
There are several test cases for the MessageRenderer class. Review them and work out which test design principles they do or don't follow.

# Acknowledgements
This example is taken from the book "Unit Testing Principles, Practices and Patterns" by Vladimir Khorikov

Reference: https://verity-ag.atlassian.net/wiki/spaces/EDU/pages/2383446024/Session+4+-+Four+pillars+of+a+good+test