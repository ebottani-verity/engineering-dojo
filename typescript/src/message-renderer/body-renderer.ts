import { Message, Renderer } from "./renderer.model";

export class BodyRenderer implements Renderer {
  render(message: Message): string {
    return `<b>${message.body}</b>`;
  }
}
