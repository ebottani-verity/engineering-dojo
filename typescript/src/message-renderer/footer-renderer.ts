import { Message, Renderer } from "./renderer.model";

export class FooterRenderer implements Renderer {
  render(message: Message): string {
    return `<i>${message.footer}</i>`;
  }
}
