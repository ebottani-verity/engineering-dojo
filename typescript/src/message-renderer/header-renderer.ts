import { Message, Renderer } from "./renderer.model";

export class HeaderRenderer implements Renderer {
  render(message: Message): string {
    return `<h1>${message.header}</h1>`;
  }
}
