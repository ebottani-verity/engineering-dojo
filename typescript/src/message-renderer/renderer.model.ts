export interface Message {
  header: string;
  body: string;
  footer: string;
}

export interface Renderer {
  render(message: Message): string;
}
