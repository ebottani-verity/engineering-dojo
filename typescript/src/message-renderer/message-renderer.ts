import { Message, Renderer } from "./renderer.model";
import { HeaderRenderer } from "./header-renderer";
import { BodyRenderer } from "./body-renderer";
import { FooterRenderer } from "./footer-renderer";

export class MessageRenderer implements Renderer {
  private subRenderers: Renderer[];

  constructor() {
    this.subRenderers = [
      new HeaderRenderer(),
      new BodyRenderer(),
      new FooterRenderer(),
    ];
  }

  public getSubRenderers(): Renderer[] {
    return this.subRenderers;
  }

  public render(message: Message): string {
    return this.subRenderers
      .map((renderer) => renderer.render(message))
      .reduce((acc, renderedPart) => acc + renderedPart, "");
  }
}
