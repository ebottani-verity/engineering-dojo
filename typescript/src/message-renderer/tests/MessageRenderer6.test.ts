import { Message } from "../renderer.model";
import { MessageRenderer } from "../message-renderer";

describe("MessageRenderer", () => {
  test("message_renderer", () => {
    const message: Message = {
      header: "header",
      body: "body",
      footer: "footer",
    };
    const sut = new MessageRenderer();

    const html = sut.render(message);

    expect(html).toMatchSnapshot();
  });
});
