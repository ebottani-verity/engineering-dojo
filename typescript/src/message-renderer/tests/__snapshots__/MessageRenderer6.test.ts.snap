// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageRenderer message_renderer 1`] = `"<h1>header</h1><b>body</b><i>footer</i>"`;
